.github/workflows/main.yml

```diff
@@ -30,3 +30,67 @@ jobs:
         run: python -m pip install -U tox
       - name: Run lint
         run: tox -elint
+  benchmark:
+    runs-on: ubuntu-latest
+    name: Run changed benchmarks
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 2
+      - name: Set up Python 3.10
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.10"
+      - name: Install Deps
+        run: |
+          python -m pip install --upgrade pip
+          pip install .
+      - name: Get changes in applications
+        id: changes-applications
+        uses: tj-actions/changed-files@v24.1
+        with:
+          files: |
+            red_queen/games/applications/*py
+      - name: Test applications
+        run: |
+          if [[ "${{steps.changes-applications.outputs.all_changed_and_modified_files}}" != "" ]]; then
+            if [[ "${{steps.changes-applications.outputs.all_changed_and_modified_files}}" == *"red_queen/games/applications/__init__.py"* ]]; then
+                no_init=${${{steps.changes-applications.outputs.all_changed_and_modified_files}}//"red_queen/games/applications/__init__.py"/}
+                echo "$no_init"
+                if [[ "$no_init" == "" ]]; then
+                    pytest red_queen/games/applications/run_bv.py -n 2 -k "not backend0 and not backend1 and not backend2" --verbose --store
+                else
+                    pytest "$no_init" -n 2 -k "not backend0 and not backend1 and not backend2" --verbose --store
+                fi
+            else
+              pytest ${{steps.changes-applications.outputs.all_changed_and_modified_files}} -n 2 -k "not backend0 and not backend1 and not backend2" --verbose --store
+            fi
+          fi 
+      - name: Get changes in mapping
+        id: changes-mapping
+        uses: tj-actions/changed-files@v24.1
+        with:
+          files: |
+            red_queen/games/mapping/*py
+      - name: Test mapping
+        run: |
+          if [[ "${{steps.changes-mapping.outputs.all_changed_and_modified_files}}" != "" ]]; then
+            if [[ "${{steps.changes-mapping.outputs.all_changed_and_modified_files}}" == *"red_queen/games/mapping/__init__.py"* ]]; then
+                no_init=${${{steps.changes-mapping.outputs.all_changed_and_modified_files}}//"red_queen/games/mapping/__init__.py"/}
+                echo "$no_init"
+                if [[ "$no_init" == "" ]]; then
+                    pytest red_queen/games/mapping/map_misc.py -n 2 -k "not backend0 and not backend1 and not backend2" --verbose --store
+                else
+                    pytest "$no_init" -n 2 -k "not backend0 and not backend1 and not backend2" --verbose --store
+                fi
+            else
+              pytest ${{steps.changes-mapping.outputs.all_changed_and_modified_files}} -n 2 -k "not backend0 and not backend1 and not backend2" --verbose --store
+            fi
+          fi 
```

Comment:
I know we talked about how to do this in a single for loop by calling git directly. But another thought which I think will be easier to implement is to abstract this into a custom workflow directive. Since this is parameterized on the path `red_queen/games/*/` we can create a custom action that takes this as a single parameter, expand it where necessary and then run the logic. Then this block would become something like:

```yaml
- name: Run changed application benchmarks:
  uses: run-changed-benchmarks
  with:
      path: red_queen/games/applications
      default_bench: red_queen/games/applications/run_bv.py
- name: Run changed mapping benchmarks:
  uses: run-changed-benchmarks
  with:
      path: red_queen/games/mapping
      default_bench:  red_queen/games/mapping/map_misc.py
```